cpp/src/parquet/variant.cc

```diff
@@ -0,0 +1,744 @@
+  std::stringstream ss;
+  ss << "ObjectInfo{"
+     << "num_elements=" << num_elements
+     << ", id_size=" << static_cast<int>(id_size)
+     << ", offset_size=" << static_cast<int>(offset_size)
+     << ", id_start_offset=" << id_start_offset
+     << ", offset_start_offset=" << offset_start_offset
+     << ", data_start_offset=" << data_start_offset
+     << "}";
+  return ss.str();
+}
+
+
+VariantValue::ObjectInfo VariantValue::getObjectInfo() const {
+  checkBasicType(VariantBasicType::Object);
+  uint8_t value_header = value[0] >> 2;
+  uint8_t field_offset_size = (value_header & 0b11) + 1;
+  uint8_t field_id_size = ((value_header >> 2) & 0b11) + 1;
+  bool is_large = ((value_header >> 4) & 0b1);
+  uint8_t num_elements_size = is_large ? 4 : 1;
+  if (value.size() < 1 + num_elements_size) {
+    throw ParquetException("Invalid object value: too short: " +
+                           std::to_string(value.size()) + " for at least " +
+                           std::to_string(1 + num_elements_size));
+  }
+  // parse num_elements
+  uint32_t num_elements = 0;
+  {
+    memcpy(&num_elements, value.data() + 1, num_elements_size);
+    num_elements = arrow::bit_util::FromLittleEndian(num_elements);
+  }
+  ObjectInfo info{};
+  info.num_elements = num_elements;
+  info.id_size = field_id_size;
+  info.offset_size = field_offset_size;
+  info.id_start_offset = 1 + num_elements_size;
+  info.offset_start_offset = info.id_start_offset + num_elements * field_id_size;
+  info.data_start_offset = info.offset_start_offset + (num_elements + 1) * field_offset_size;
+  // Check the boundary with the final offset
+  if (info.data_start_offset > value.size()) {
+    throw ParquetException("Invalid object value: data_start_offset=" +
+                                 std::to_string(info.data_start_offset) +
+                                 ", value_size=" + std::to_string(value.size()));
+  }
+  {
+    uint32_t final_offset = 0;
+    memcpy(&final_offset,
+           value.data() + info.offset_start_offset + num_elements * field_offset_size,
+           field_offset_size);
+    // It could be less than value size since it could be a sub-object.
+    if (final_offset + info.data_start_offset > value.size()) {
+      throw ParquetException("Invalid object value: final_offset=" +
+                             std::to_string(final_offset) +
+                             ", data_start_offset=" + std::to_string(info.data_start_offset) +
+                             ", value_size=" + std::to_string(value.size()));
+    }
+  }
+  return info;
+}
+
+std::optional<VariantValue> VariantValue::getObjectValueByKey(
+    std::string_view key) const {
+  ObjectInfo info = getObjectInfo();
+
+  return getObjectValueByKey(key, info);
+}
+
+std::optional<VariantValue> VariantValue::getObjectValueByKey(
+    std::string_view key, const VariantValue::ObjectInfo& info) const {
+  // TODO(mwish): Currently we just linear search here. The best way here is:
```

Comment:
Currently I just do linear search, this could be optimized later, but I don't want to finish it in this pr (will create an github issue when everyone agree with this interface)